src/Native/CpuMathNative/Sse.cpp

```diff
@@ -1864,35 +1880,102 @@ EXPORT_API(void) MulElementWiseSU(_In_ const float * ps1, _In_ const float * ps2
         pd[*pi] = ps1[*pi] * ps2[*pi];    
 }
 
-EXPORT_API(float) SumA(const float * ps, int c)
+EXPORT_API(float) Sum(const float* pValues, int length)
 {
-    const float * psLim = ps + c;
+    if (length < 4)
+    {
+        // Handle cases where we have less than 128-bits total and can't ever use SIMD acceleration.
 
-    __m128 res = _mm_setzero_ps();
-    for (; ps < psLim; ps += 4)
-        res = _mm_add_ps(res, _mm_load_ps(ps));
+        float result = 0;
 
-    res = _mm_hadd_ps(res, res);
-    res = _mm_hadd_ps(res, res);
+        switch (length)
+        {
+            case 3: result += pValues[2];
+            case 2: result += pValues[1];
+            case 1: result += pValues[0];
+        }
 
-    return _mm_cvtss_f32(res);
-}
+        return result;
+    }
 
-EXPORT_API(float) SumU(const float * ps, int c)
-{
-    const float * psLim = ps + c;
+    __m128 result = _mm_setzero_ps();
 
-    __m128 res = _mm_setzero_ps();
-    for (; ps + 4 <= psLim; ps += 4)
-        res = _mm_add_ps(res, _mm_loadu_ps(ps));
+    uintptr_t address = (uintptr_t)(pValues);
+    uintptr_t misalignment = address % 16;
 
-    res = _mm_hadd_ps(res, res);
-    res = _mm_hadd_ps(res, res);
+    int remainder = 0;
 
-    for (; ps < psLim; ps++)
-        res = _mm_add_ss(res, _mm_load_ss(ps));
+    if ((misalignment & 3) != 0)
+    {
+        // Handles cases where the data is not 32-bit aligned and we can't ever use aligned operations
```

Comment:
I'm not following this reasoning. Why can't we do scalar operations in this case in order to get aligned?

Reply:
Ah, I get it now. All the `float`s in the array are not aligned on a `float` boundary. Seems like an odd situation, but I understand it now.